File: Python/programmers/etc/level3/다단계_칫솔_판매.py
def solution(enroll, referral, seller, amount):
    graph, ans = {}, {e:0 for e in enroll}

    for e, r in zip(enroll, referral):
        graph[e] = r

    for s, a in zip(seller, amount):
        money = a*100
        rate = money//10
        ans[s] += money-rate
        x = graph[s]

        while x != "-":
            if rate == 0: 
                break
            ans[x] += rate - rate//10
            rate //= 10
            x = graph[x]

    return list(ans.values())


enroll = ["john", "mary", "edward", "sam", "emily", "jaimie", "tod", "young"]
referral = ["-", "-", "mary", "edward", "mary", "mary", "jaimie", "edward"]
seller = ["young", "john", "tod", "emily", "mary"]
amount = [12, 4, 2, 5, 10]
print(solution(enroll, referral, seller, amount)) # [360, 958, 108, 0, 450, 18, 180, 1080]

enroll = ["john", "mary", "edward", "sam", "emily", "jaimie", "tod", "young"]
referral = ["-", "-", "mary", "edward", "mary", "mary", "jaimie", "edward"]
seller = ["sam", "emily", "jaimie", "edward"]
amount = [2, 3, 5, 4]
print(solution(enroll, referral, seller, amount)) # [0, 110, 378, 180, 270, 450, 0, 0]

File: Java/programmers/etc/level3/다단계칫솔판매.java
package CodeTest.Java.programmers.etc.level3;

import java.util.*;

public class 다단계칫솔판매 {
    public int[] solution(String[] enroll, String[] referral, String[] seller, int[] amount) {
        Map<String, Integer> map = new HashMap<>();
        Map<String, String> graph = new HashMap<>();

        int n = enroll.length;
        for (int i=0; i<n; i++) {
            map.put(enroll[i], 0);
            graph.put(enroll[i], referral[i]);
        }

        int m = seller.length;
        for (int i=0; i<m; i++) {
            String s = seller[i];
            int a = amount[i];

            int money = a*100;
            int rate = money/10;
            map.put(s, map.get(s)+(money-rate));
            String x = graph.get(s);

            while (!x.equals("-")) {
                if (rate == 0) break;
                map.put(x, map.get(x)+(rate-rate/10));
                rate /= 10;
                x = graph.get(x);
            }
        }

        int[] answer = new int[n];
        for (int i=0; i<n; i++) {
            answer[i] = map.get(enroll[i]);
        }

        return answer;
    }
}
